File: zul/src/com/potix/zul/html/Treecell.java
/* Treecell.java

{{IS_NOTE
	$Id: Treecell.java,v 1.18 2006/03/17 10:06:38 tomyeh Exp $
	Purpose:
		
	Description:
		
	History:
		Wed Jul  6 18:56:30     2005, Created by tomyeh@potix.com
}}IS_NOTE

Copyright (C) 2005 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
	This program is distributed under GPL Version 2.0 in the hope that
	it will be useful, but WITHOUT ANY WARRANTY.
}}IS_RIGHT
*/
package com.potix.zul.html;

import java.util.List;
import java.util.LinkedList;
import java.util.Iterator;

import com.potix.xml.HTMLs;

import com.potix.zk.ui.Component;
import com.potix.zk.ui.UiException;

import com.potix.zul.html.impl.LabelImageElement;

/**
 * A treecell.
 *
 * <p>In XUL, treecell cannot have any child, but ZUL allows it.
 * Thus, you could place any kind of children in it. They will be placed
 * right after the image and label.
 *
 * @author <a href="mailto:tomyeh@potix.com">tomyeh@potix.com</a>
 * @version $Revision: 1.18 $ $Date: 2006/03/17 10:06:38 $
 */
public class Treecell extends LabelImageElement {
	private static final String ROOT_OPEN = "/root-open.gif";
	private static final String ROOT_CLOSE = "/root-close.gif";
	private static final String LAST_OPEN = "/last-open.gif";
	private static final String LAST_CLOSE = "/last-close.gif";
	private static final String TEE_OPEN = "/tee-open.gif";
	private static final String TEE_CLOSE = "/tee-close.gif";
	private static final String TEE = "/tee.gif";
	private static final String LAST = "/last.gif";
	private static final String VBAR = "/vbar.gif";
	private static final String SPACER = "/spacer.gif";

	public Treecell() {
	}
	public Treecell(String label) {
		setLabel(label);
	}
	public Treecell(String label, String src) {
		setLabel(label);
		setSrc(src);
	}

	/** Return the tree that owns this cell.
	 */
	public Tree getTree() {
		for (Component n = this; (n = n.getParent()) != null;)
			if (n instanceof Tree)
				return (Tree)n;
		return null;
	}
	/** Returns the tree col associated with this cell, or null if not available.
	 */
	public Treecol getTreecol() {
		final Tree tree = getTree();
		if (tree != null) {
			final Treecols lcs = tree.getTreecols();
			if (lcs != null) {
				final int j = getColumnIndex();
				final List lcschs = lcs.getChildren();
				if (j < lcschs.size())
					return (Treecol)lcschs.get(j);
			}
		}
		return null;
	}
	/** Returns the column index of this cell, starting from 0.
	 */
	public int getColumnIndex() {
		int j = 0;
		for (Iterator it = getParent().getChildren().iterator();
		it.hasNext(); ++j)
			if (it.next() == this)
				break;
		return j;
	}

	/** Returns the maximal length for this cell, which is decided by
	 * the corresponding {@link #getTreecol}'s {@link Treecol#getMaxlength}.
	 *
	 * <p>Note: DBCS counts  two bytes (range 0x4E00~0x9FF).
	 * Default: 0 (no limit).
	 */
	public int getMaxlength() {
		final Tree tree = getTree();
		if (tree == null)
			return 0;
		final Treecol lc = getTreecol();
		return lc != null ? lc.getMaxlength(): 0;
	}

	/** Returns the level this cell is. The root is level 0.
	 */
	public int getLevel() {
		final Component parent = getParent();
		return parent != null ? ((Treerow)parent).getLevel(): 0;
	}

	//-- Internal use only --//
	/** Returns the prefix of the first column (in HTML tags), null if this
	 * is not first column. Called only by treecell.jsp.
	 */
	public String getColumnHtmlPrefix() {
		if (isFirstColumn()) {
			final Treeitem item = getTreeitem();
			final Tree tree = getTree();
			final StringBuffer sb = new StringBuffer(80);
			if (tree != null && tree.isCheckmark()) {
				sb.append("<input type=\"")
					.append(tree.isMultiple() ? "checkbox": "radio")
					.append('"');
				if (item.isSelected())
					sb.append(" checked=\"checked\"");
				sb.append(" id=\"")
					.append(getParent().getUuid()).append("!cm\"")
					.append(" zk_type=\"Tcfc\"").append("/>");
					//NOTE: use Treerow's uuid! NOT Treeitem's!
			}

			final Treeitem[] pitems = getTreeitems(item);
			for (int j = 0; j < pitems.length; ++j)
				appendImage(sb,
					j == 0 || isLastChild(pitems[j]) ? SPACER: VBAR, false);
	
			if (item.isContainer()) {
				appendImage(sb,
					item.isOpen() ?
						pitems.length == 0 ? ROOT_OPEN:
							isLastChild(item) ? LAST_OPEN: TEE_OPEN:
						pitems.length == 0 ? ROOT_CLOSE:
							isLastChild(item) ? LAST_CLOSE: TEE_CLOSE,
						true);
			} else {
				appendImage(sb,
					pitems.length == 0 ? SPACER:
						isLastChild(item) ? LAST: TEE, false);
			}
			return sb.toString();
		} else {
			//To make the tree's height more correct, we have to generate &nbsp;
			//for empty cell. Otherwise, IE will make the height too small
			final boolean empty = getImage() == null && getLabel().length() == 0
				&& getChildren().isEmpty();
			return empty ? "&nbsp;": null;
		}
	}
	/** Returns the postfix of the first column (in HTML tags), null if this
	 * is not first column. Called only by treecell.jsp.
	 */
	public String getColumnHtmlPostfix() {
		final Treeitem item = getTreeitem();
		if (isFirstColumn() && item.isFocusRequired()) {
			final StringBuffer sb = new StringBuffer(64)
				.append("<a href=\"javascript:;\" id=\"")
				.append(getParent().getUuid()).append("!sel\"")
				.append(" zk_type=\"Tcfc\"").append("> </a>");
					//NOTE: use Treerow's uuid! NOT Treeitem's!
			return sb.toString();
		} else { 
			return null;
		}
	}
	/** Returns whether this is the first column. */
	private boolean isFirstColumn() {
		final Component parent = getParent();
		return parent != null && parent.getChildren().get(0) == this;
	}

	/** Returns whether an item is the last child.
	 */
	public boolean isLastChild(Treeitem item) {
		final Component parent = item.getParent();
		if (parent == null)
			return true;

		final List sibs = parent.getChildren();
		return sibs.get(sibs.size() - 1) == item;
	}
	/** Returns an array of Treeitem from the root.
	 */
	private Treeitem[] getTreeitems(Component item) {
		final List pitems = new LinkedList();
		for (;;) {
			final Component tch = item.getParent();
			if (tch == null)
				break;
			item = tch.getParent();
			if (item == null || item instanceof Tree)
				break;
			pitems.add(0, item);
		}
		return (Treeitem[])pitems.toArray(new Treeitem[pitems.size()]);
	}

	/** Returns the nestest parent {@link Treeitem}.
	 */
	public Treeitem getTreeitem() {
		final Component parent = getParent();
		return parent != null ? (Treeitem)parent.getParent(): null;
	}

	/** Generates HTML tags for &lt;img&gt;.
	 * @param button whether this is the button to toggle open/close
	 */
	private void appendImage(StringBuffer sb, String name, boolean button) {
		final int width = getIconWidth(), height = getIconHeight();
		final String uri = getIconURI(name);
		sb.append("<img align=\"top\"");
		HTMLs.appendAttribute(
			sb, "src", getDesktop().getExecution().encodeURL(uri));
		HTMLs.appendAttribute(sb, "width", width);
		HTMLs.appendAttribute(sb, "height", height);
		if (button) {
			final Treeitem item = getTreeitem();
			if (item != null) {
				sb.append(" zk_type=\"Tcop\"");
				HTMLs.appendAttribute(sb,
					"id", item.getTreerow().getUuid() + "!open");
			}

			//HTMLs.appendAttribute(sb, "title", title);
		}
		sb.append("/>");
	}

	/** Returns the width of the icon. */
	private int getIconWidth() {
		return getRequiredIntInitParam("icon-width");
	}
	/** Returns the height of the icon. */
	private int getIconHeight() {
		return getRequiredIntInitParam("icon-height");
	}
	/** Returns the icon URI.
	 */
	private String getIconURI(String name) {
		final String s = getInitParam("icon-uri");
		if (s == null)
			throw new UiException("The icon-uri param must be defined");
		return s + name;
	}
	private int getRequiredIntInitParam(String name) {
		final int v = getIntInitParam(name);
		if (v <= 0)
			throw new UiException("The "+name+" param must be defined and positive");
		return v;
	}

	//-- super --//
	/** Returns the width which the same as {@link #getTreecol}'s width.
	 */
	public String getWidth() {
		final Treecol col = getTreecol();
		return col != null ? col.getWidth(): null;
	}
	public void setWidth(String width) {
		throw new UnsupportedOperationException("Set treecol's width instead");
	}
	public String getOuterAttrs() {
		final Treecol col = getTreecol();
		final String attrs = super.getOuterAttrs();
		return col != null ? attrs+col.getColAttrs(): attrs;
	}

	//-- Component --//
	public void setParent(Component parent) {
		if (parent != null && !(parent instanceof Treerow))
			throw new UiException("Wrong parent: "+parent);
		super.setParent(parent);
	}
	public void invalidate(Range range) {
		super.invalidate(range);
		initAtClient();
	}
	public void onChildAdded(Component child) {
		super.onChildAdded(child);
		initAtClient();
	}
	public void onChildRemoved(Component child) {
		super.onChildRemoved(child);
		initAtClient();
	}
	private void initAtClient() {
		final Tree tree = getTree();
		if (tree != null) tree.initAtClient();
	}
}

File: zul/src/archive/web/js/zul/sel/Treecell.js
/* Treecell.js

	Purpose:
		
	Description:
		
	History:
		Wed Jun 10 15:32:39     2009, Created by jumperchen

Copyright (C) 2009 Potix Corporation. All Rights Reserved.

This program is distributed under LGPL Version 2.0 in the hope that
it will be useful, but WITHOUT ANY WARRANTY.
*/
(function () {
/**
 * A treecell.
 *
 * <p>In XUL, treecell cannot have any child, but ZUL allows it.
 * Thus, you could place any kind of children in it. They will be placed
 * right after the image and label.
 *
 * <p>Default {@link #getZclass}: z-treecell
 */
zul.sel.Treecell = zk.$extends(zul.LabelImageWidget, {
	/**
	 * The width can't be specified in this component.
	 */
	setWidth: zk.$void, // readonly
	_colspan: 1,
	$define: {
    	/** Returns number of columns to span this cell.
    	 * Default: 1.
    	 * @return int
    	 */
    	/** Sets the number of columns to span this cell.
    	 * <p>It is the same as the colspan attribute of HTML TD tag.
    	 * @param int colspan
    	 */
		colspan: [
			function (colspan) {
				return colspan > 1 ? colspan: 1;
			},
			function () {
				var n = this.$n();
				if (n) n.colSpan = this._colspan;
			}]
	},
	/** Return the tree that owns this cell.
	 * @return Tree
	 */
	getTree: function () {
		return this.parent ? this.parent.getTree() : null;
	},
	domStyle_: function (no) {
		var style = this.$super('domStyle_', zk.copy(no, {width:true})),
				//bug#3185657: not span content if given width
			tc = this.getTreecol();
			// B70-ZK-2946: adds the text-align from treecol, same as Listcell
			if (tc) if (tc._align) style += 'text-align: ' + tc._align + ';';
			return this.isVisible() && tc && !tc.isVisible() ? style +
				'display:none;' : style;
	},
	/** Returns the tree col associated with this cell, or null if not available.
	 * @return Treecol
	 */
	getTreecol: function () {
		var tree = this.getTree();
		if (tree && tree.treecols) {
			var j = this.getChildIndex();
			if (j < tree.treecols.nChildren)
				return tree.treecols.getChildAt(j);
		}
		return null;
	},
	/** Returns the level this cell is. The root is level 0.
	 * @return int
	 */
	getLevel: function () {
		return this.parent ? this.parent.getLevel(): 0;
	},
	/** Returns the maximal length of each item's label.
	 * @return int
	 */
	getMaxlength: function () {
		var tc = this.getTreecol();
		return tc ? tc.getMaxlength() : 0;
	},
	domLabel_: function () {
		return zUtl.encodeXML(this.getLabel(), {maxlength: this.getMaxlength()});
	},
	getTextNode: function () {
		return this.getCaveNode();
	},
	domContent_: function () {
		var s1 = this.$supers('domContent_', arguments),
			s2 = this._colHtmlPre();
		return s1 ? s2 ? s2 + '<span class="' + this.$s('text') + '">&nbsp;' + s1 + '</span>' : s1: s2;
	},
	bind_: function () {
		this.$supers('bind_', arguments);
		if (this._clearCache) { // B60-ZK-1348
			this._clearCache = false;
			var p;
			if (p = this.parent) {
				p.clearCache(); //$n('open')
			}
		}
	},
	doMouseOver_: function(evt) {
		var n = this.$n();
		
		// ZK-2136: all children should apply -moz-user-select: none
		if (n && zk.gecko && (this._draggable || this.parent._draggable)
				&& !jq.nodeName(evt.domTarget, 'input', 'textarea')) {
			jq(n).addClass('z-draggable-over');
		}
		this.$supers('doMouseOver_', arguments);
	},
	doMouseOut_: function(evt) {
		var n = this.$n();
		
		// ZK-2136: all children should apply -moz-user-select: none
		if (n && zk.gecko && (this._draggable || this.parent._draggable)
				&& !jq.nodeName(evt.domTarget, 'input', 'textarea')) {
			jq(n).removeClass('z-draggable-over'); // Bug ZK-580
		}
		this.$supers('doMouseOut_', arguments);
	},
	doFocus_: function (evt) {
		this.$supers('doFocus_', arguments);
		//sync frozen
		var tree = this.getTree(),
		frozen = tree ? tree.frozen : null,
		tbody = tree && tree._treechildren ? tree._treechildren.$n() : null,
		td, tds;
		if (frozen && tbody) {
			tds = jq(evt.domTarget).parents('td');
			for (var i = 0, j = tds.length; i < j; i++) {
				td = tds[i];
				if (td.parentNode.parentNode == tbody) {
					tree._moveToHidingFocusCell(td.cellIndex);
					break;
				}
			}
		}
	},
	_syncIcon: function (isRemoved) {
		this.rerender(isRemoved ? -1 : false);
		var p;
		if (p = this.parent) {
			this._clearCache = true;
		}
	},
	_colHtmlPre: function () {
		if (this.parent.firstChild == this) {
			var item = this.parent.parent,
				tree = item.getTree(),
				sb = new zk.Buffer();
			if (tree) {
				if (tree.isCheckmark()) {
					var chkable = item.isSelectable(),
						multi = tree.isMultiple(),
						cmCls = multi ? item.$s('checkbox') : item.$s('radio'),
						ckCls = multi ? ' z-icon-check' : ' z-icon-radio';
					sb.push('<span id="', this.parent.uuid, '-cm" class="',
							item.$s('checkable'), ' ', cmCls);
					
					if (!chkable || item.isDisabled())
						sb.push(' ', item.$s('disabled'));
					
					sb.push('"');
					if (!chkable)
						sb.push(' style="visibility:hidden"');
					
					sb.push('><i class="', item.$s('icon'), ckCls, '"></i></span>');
				}
			}
			var iconScls = tree ? tree.getZclass() : '',
				pitems = this._getTreeitems(item, tree);
			for (var j = 0, k = pitems.length; j < k; ++j)
				this._appendIcon(sb, iconScls, 'spacer', false);
			
			if (item.isContainer()) {
				var name = item.isOpen() ? 'open' : 'close';
				this._appendIcon(sb, iconScls, name, true);
			} else {
				this._appendIcon(sb, iconScls, 'spacer', false);
			}
			return sb.join('');
		} else {
			//To make the tree's height more correct, we have to generate &nbsp;
			//for empty cell. Otherwise, IE will make the height too small
			return !this.getImage() && !this.getLabel()	&& !this.nChildren ? '&nbsp;' : null;
		}
	},
	_getTreeitems: function (item, tree) {
		var pitems = [];
		for (;;) {
			var tch = item.parent;
			if (!tch)
				break;
			item = tch.parent;
			if (!item || item == tree)
				break;
			pitems.unshift(item);
		}
		return pitems;
	},
	_appendIcon: function (sb, iconScls, name, button) {
		var openCloseIcon = '';
		sb.push('<span class="');
		if (name == 'spacer') {
			sb.push(iconScls, '-line ', iconScls, '-', name, '"');
			openCloseIcon += '&nbsp;';
		} else {
			var id = '';
			if (button) {
				var item = this.parent;
				if (item)
					id = item.uuid + '-icon';
			}
			sb.push(iconScls, '-icon"');
			var icon = this.getIconOpenClass_();
			if (name.indexOf('close') > -1)
				icon = this.getIconCloseClass_();
			
			openCloseIcon += '<i id="' + id + '" class="' + icon + ' ' + iconScls +
					'-' + name + '"></i>';
		}
		if (button) {
			var item = this.parent; // B65-ZK-1608, appendChild() will invoke before treeitem._fixOnAdd() 
			if (item)
				sb.push(' id="', item.uuid, '-open"');
		}
		sb.push('>', openCloseIcon, '</span>');
		openCloseIcon = null;
	},
	getIconOpenClass_: function () {
		return 'z-icon-caret-down';
	},
	getIconCloseClass_: function () {
		return 'z-icon-caret-right';
	},
	getWidth: function() {
		var col = this.getTreecol();
		return col ? col.getWidth() : null;
	},
	domAttrs_: function () {
		return this.$supers('domAttrs_', arguments)
			+ (this._colspan > 1 ? ' colspan="' + this._colspan + '"' : '');
	},
	updateDomContent_: function () {
		this.$supers('updateDomContent_', arguments);
		if (this.parent)
			this.parent.clearCache();
	},
	deferRedrawHTML_: function (out) {
		out.push('<td', this.domAttrs_({domClass:1}), ' class="z-renderdefer"></td>');
	}
});
})();